Document phase zinc

karsmars opened the pull request.

karsmars committed:
Update README

karsmars committed:
Added Overview to README

karsmars committed:
Brief overview of files included in phase_zinc

karsmars committed:
Update hyperlink in README

karsmars committed:
Add documentation explaining phase architecture

karsmars committed:
Document phase_zinc_compile_ext.bzl

karsmars committed:
Trace Scala/Zinc config, Zinc compile runner

karsmars committed:
Explain drawbacks of Zinc

karsmars committed:
Document phase_zinc

karsmars committed:
Add drawbacks to README

karsmars committed:
Update README internal links

karsmars committed:
Update README: how to set up phase_zinc

karsmars committed:
Update setup instructions in README

karsmars committed:
Update README for clarity

karsmars committed:
Explain where "@zinc" workspace comes from

karsmars committed:
Add src directory to README

borkaehw commented on README.md:
```diff
@@ -1 +1,55 @@
+
+The [src/main/scala](src/main/scala) directory contains Scala and Zinc configuration data. See [src/main/scala/workers/BUILD.bazel](src/main/scala/workers/BUILD.bazel).
+
+The [scripts](scripts) directory contains shell scripts to run these tests. Run `scripts/travis.sh` to execute all tests.
```
I don't think it's necessary. Users shouldn't care about how to run tests nor how we set up the CI. Plus, if we want to merge it back to `bazelbuild/rules_scala`, we will discard all scripts. We should use their scripts.

karsmars replied:
As far as I know, our current goal is not to merge back into `bazelbuild/rules_scala`. We just want to add an external link from `bazelbuild/rules_scala` to this repository.

karsmars replied:
I also thought that having explanations for the different directories will help first-time users know what is going on.

borkaehw commented on README.md:
```diff
@@ -1 +1,55 @@
+This adds the `phase_zinc` repo to your workspace and loads some dependencies. Make sure you have the desired commit hash (in case this repository is updated but this README is not). Also check to make sure the sha256 matches.
+
+
+Sweet, now you can use `zinc_scala_binary` as a normal rule! Just add the following line to any `BUILD` file:
```
Again, we should instruct the users the correct way to use phase_zinc. Using `zinc_scala_binary` loses the ability to add other custom phases.

karsmars replied:
Ah, I understand what you mean. I made some edits, let me know what you think

borkaehw commented on rules/ext/phase_zinc_compile_ext.bzl:
```diff
@@ -26,9 +33,13 @@ ext_zinc_compile = {
 def _zinc_compile_singleton_implementation(ctx):
     return [
         _ScalaRulePhase(
+            # When the maker_scala_binary (or similar) function runs, it will assign the label "compile" to _phase_zinc_compile
```
Should be `make_scala_binary`.

borkaehw commented on rules/scala.bzl:
```diff
@@ -14,21 +17,30 @@ load(
     _ZincInfo = "ZincInfo",
 )
+
+# This line is unneeded, as phase_zinc_compile is already wrapped inside ext_zinc_compile (imported below)
```
If it's unused, let's remove it with a separate PR, rather than documenting it this way.

borkaehw reviewed (changes requested):
Looks great in general. Your language is very easy to understand, great job on explaining all these. Left a few comments.

borkaehw commented:
Since `Add ZincConfiguration to list of providers` is on `master`, could you rebase it so we don't see the diff from that commit in this PR? Thanks.

karsmars commented on README.md:
```diff
@@ -1 +1,55 @@
+
+## Files included
+
+Phase_zinc defines a Zinc compiler phase, then adds this phase to three Bazel rules originally defined in [bazelbuild/rules_scala](https://github.com/bazelbuild/rules_scala): `scala_binary`, `scala_library` and `scala_test`. The new corresponding rules are named, respectively, `zinc_scala_binary`, `zinc_scala_library` and `zinc_scala_test`.
```
In `rules/scala.bzl`, we specifically import make functions for `scala_binary`, etc, and then we implement `zinc_scala_binary` in the same file. Are you suggesting we remove that from `scala.bzl`?

karsmars replied:
I drew inspiration from https://github.com/bazelbuild/rules_scala/blob/master/docs/phase_scalafmt.md, but I also noted that `zinc_scala_binary`, etc, are already defined in `scala.bzl` so that's why I wrote it this way. Do you think we should do it more in the style of the the `phase_scalafmt` doc?

karsmars committed:
Update README to explain how to make custom rule

Previously, the README explained directly how to use the zinc_scala_binary rule defined in rules/scala.bzl. However, we don't want to encourage this, and instead should just explain how to insert the phase into a rule macro to create a custom rule.

karsmars committed:
Specify from where files were borrowed

karsmars committed:
Explain usage of rule macros

karsmars committed:
Explain need to import `rules_scala` to WORKSPACE

borkaehw commented on README.md:
````diff
@@ -1 +1,52 @@
+```
+This adds the `phase_zinc` repo to your workspace and loads some dependencies. Make sure you have the desired commit hash (in case this repository is updated but this README is not). Also check to make sure the sha256 matches.
+
+To add this phase to a rule, you have to pass the extension to a rule macro. Take `scala_binary` for example: first, load the rule macro from [bazelbuild/rules_scala](https://github.com/bazelbuild/rules_scala), then load the extension from @phase_zinc and finally pass the extension to the macro to define a custom rule. You'll also need to make sure you have loaded the [bazelbuild/rules_scala](htt ...
````
Adding backtick to @phase_zinc would be better.

borkaehw reviewed (changes requested):
Getting very close, great job.

karsmars committed:
Update README

borkaehw reviewed (approved):
LGTM, thanks.
